### ISSUE TITLE
Portrait image being cropped to smaller 4:3 piece even with autoSquareCrop false (v3.0.0alpha-14)

### ISSUE THREAD

**Issue by Murilo-Perrone** [triniwiz/fancycamera#12]
This bug has been oberved in nativescript-camera-plus component ([here](https://github.com/nstudio/nativescript-camera-plus/issues/143)). Up to 3.0.8 pictures were ok, but in 3.1.0 pictures got smaller, because portrait mode was not detected by fancycamera.

![](https://user-images.githubusercontent.com/26028390/101922743-3df85d00-3bad-11eb-8261-a554aa2089f1.png)

As noted by @triniwiz , one reason is that mapping of `FancyCamera.cameraOrientation` property to internal camera (cameraView) is missing. But fixing that wasn't enough because it avoided the image crop but brought a -90 rotated image.

Another fix that may be necessary is to change Camera2.kt so that it also rotates the image when autoSquareCrop is false. I am testing this fix already, but may need more testing guidelines.

**Comment by triniwiz** [comment]
[This](https://github.com/triniwiz/fancycamera/blob/5a1fe0a4d274774d9d0bd868de52ed8920f7a5ee/fancycamera/src/main/java/co/fitcom/fancycamera/FancyCamera.kt) should be the way to get back things working I missed this with the changes

**Comment by Murilo-Perrone** [comment]
Thanks for the pointers @triniwiz . The code you linked confirms that the image needs to be rotated in all cases, not only when `autoSquareCrop` is used.

However when `autoSquareCrop` is false, this code uses a different strategy. It uses the faster method from ImageCapture class (which provides a jpg file), then uses ExifInterface to "rotate" the image through the image's metadata (instead of rotating the actual pixels matrix). Although this is a common practice, it is not 100% compatible. For instance I have problems with Amazon Image Rekognition console when using this images. I then use imagemagick to rotate actual pixels from pictures that came from Samsung cell phones.

So perhaps we should offer both rotation strategies? I mean to allowing user to prioritize for fast results (exif) or the maximum compatibity (in-memory pixel rotation). Let me know your thoughts. I could add an optional parameter in `FancyCamera.takePhoto()`.

The fix I have just committed [here](https://github.com/Murilo-Perrone/fancycamera/commit/0047a805710545663fd91edd052aefafb006241b) does the rotation via pixels, but more testing is still needed.

**Comment by triniwiz** [comment]
I think it should be exif by default but add in the option to rotate the pixels

**Comment by Murilo-Perrone** [comment]
After thousands of pictures, I have figured it out. I took the opportunity to upgrade for December's camerax release 1.0.0-rc01 / 1.0.0-alpha20, which is quite faster.

The [camerax guide](https://developer.android.com/training/camerax/orientation-rotation) provides the best guidelines. Three things were missing in the code:

1. The `targetRotation` must be properly calculated when building of camerax components.
2. The `pictureSize` must be specified in wither Height x Width or Width X Height depending rather the app is in portrait or landscape mode.
3. Camerax must be able to detect app switching from portrait to landscape mode, so it can adjust accordingly (which isn't happening). I believe `surfaceProvider` parameter must be connected to the actual view showing the preview, instead of the `Camera2` class.

I worked out (1) and (2), while (3) is still pending. Unsatisfied requirement (2) was responsible for the image cropping. Also, unsatisfied requirement (3) is holding up preview in portrait mode, hence I have specified `pictureSize` in Height x Width always.

I also added `allowExifRotation` property, defaulting to `true`. And I am updating demo project from nativescript-camera-plus with the new API.

![pic1](https://user-images.githubusercontent.com/26028390/102616005-9bdcf580-4115-11eb-91fc-8f805ac6d8e7.png)
![pic2](https://user-images.githubusercontent.com/26028390/102616127-d777bf80-4115-11eb-9542-3fad2eca23ab.png)
![pic3](https://user-images.githubusercontent.com/26028390/102616175-e9596280-4115-11eb-9d79-66cf654bab28.png)

I and noticed the "Retake!" dialog does not support Exit rotation, because it does pixel level resampling. That can later be fixed too.

![retakeDialog](https://user-images.githubusercontent.com/26028390/102615522-d72af480-4114-11eb-863e-f8d5b515cb20.png)

Now I'll explain `targetRotation` calculation. First thing to keep in mind is that three values must be taken into account:

1a. The device's current rotation - a constantly changing value which is detected by the device's sensors
1b. The device's specific sensor rotation - a fixed reference value
1c. The application activity rotation - which may change when you rotate the device, depending rather it is locked or not

We have (a), and perhaps (b), already being taken care by `OrientationEventListener` which is setting `currentOrientation` property in the code. So we just need to use that value combined with (c). With that in place, we should not need make any further deductions of rotation after `ImageCapture.takePicture` runs. Because the callbacks already take care of rotation based on targetRotation we provided. The `ImageProxy` instance specifies the rotation angle back, for our bitmap rotation matrix, while the callback providing a saved JPG already puts Exif rotation in it automatically.

To calculate `targetRotation` I first observed the logic found in `Camera.kt`. It combines `Camera.CameraInfo.orientation` (API 9-20) with current activity's rotation. It seems a bit experimental, as it has an exception for a specific model (Nexus6):
```kotlin
private var rotationAngle = 0
private fun updateCameraDisplayOrientation(activity: Activity, cameraId: Int, camera: Camera?) {
  val info = Camera.CameraInfo()
  Camera.getCameraInfo(cameraId, info)
  var degrees = activity.windowManager.defaultDisplay.rotation * 90
  if (info.facing == Camera.CameraInfo.CAMERA_FACING_FRONT) {
    if (isNexus6) angle = 90
    else angle = (info.orientation + degrees) % 360
  } else {
    angle = (info.orientation - degrees + 360) % 360
```

For new api, I found two new properties used by `Camera2.kt`:
```kotlin
internal var currentOrientation: Int = OrientationEventListener.ORIENTATION_UNKNOWN // CameraBase.kt
override var rotation: CameraOrientation = CameraOrientation.UNKNOWN // Camera2.kt
```
- `currentOrientation`: being automatically set by `OrientationEventListener`
- `rotation`: not being set anywhere and not mapped to nativescript-camera-plus, but was being used to calculate targetRotation of camerax. I'm assuming this was unfinished work and left it alone.

So I seem to have reached proper calculation of targetRotation with this code:
```kotlin
  private fun getTargetRotation(): Int {
      val windowService = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
      @Suppress("DEPRECATION") val displayRotation: Int = windowService.getDefaultDisplay().getRotation()
      val deviceRotation = if (currentOrientation >= 0) currentOrientation / 90 else 0
      return when (position) {
          CameraPosition.BACK -> (4 + deviceRotation + displayRotation) % 4
          CameraPosition.FRONT -> (8 + deviceRotation - displayRotation) % 4
      }
  }
```

Now with (1) and (2) taken care, all combinations of use case can produced the same image as shown in preview - back or front camera, all 4 orientations, with app in landscape or portrait mode. Yet since (3) wasn't fixed, preview is not OK when app turns. This can be observed in with nativescript-camera-plus demo app, when it tilts to landscape mode. The preview window does not turn accordingly.

So, meanwhile I added code in `ImageCapture.takePicture` callbacks to compensate for the undetected tilt. Though I know that this is temporary, the code "chasing its own tail", paying a price for the incorrect building of the camerax components. When (3) gets fixed we can get rid of `fixImageProxyTargetRotation` and `fixJpgImageExifRotation` methods that I created, and also uncomment the code I added to `pictureSize` property.

There is a project which claims to have no issues, named [arindamxd/android-camerax](https://github.com/arindamxd/android-camerax/blob/2408cbc847246d6612bf8262eb9b6c81592cafb2/app/src/main/java/com/arindam/camerax/ui/home/camera/CameraFragment.kt#L340).

Please let me know if there is anything I need to know before I do a final commit and pull request. I believe it is a release candidate already.

**Comment by Murilo-Perrone** [comment]
I noticed Camera2 components are not being rebuilt until preview is re-started. Thus the fixes are only being effective after:
```TS
onOrientationChanged(eventName: string, newValue: string) {
  if (isAndroid && this.cam) {
    const fancyCamera = (<any>this.cam)._camera as com.github.triniwiz.fancycamera.FancyCamera;
    fancyCamera.stopPreview();
    setTimeout(() => fancyCamera.startPreview(), 1000);
  }
```
In my tests I used this trigger as a workaround:
```TS
constructor(...) {
  on("orientationChanged",  (evt) => {
    setTimeout(() => this.onOrientationChanged(evt.eventName, evt.newValue), 1000) // Restart preview
  });
```
Now we need to find out which is the best way to communicate the view updates to `Camera2.kt`, overlooking compatibility with nativescript-camera-plus...

**Comment by Murilo-Perrone** [comment]
fixed by #13
